File: data-structures/java/Queue.java
public class Queue<T>
{
    private int size = 0;
    private T defaultValue;
    private SingleLinkNode<T> back = null;
    private SingleLinkNode<T> front = null;

    public Queue() {}

    public void enqueue(T data)
    {
        this.size++;
        SingleLinkNode<T> node = new SingleLinkNode<T>(data);

        if(this.back == null || this.front == null) { this.back = node; this.front = node; }
        else { this.back.setNext(node); this.back = node; }
    }

    public void dequeue()
    {
        if(this.front == null) return;

        this.size--;
        SingleLinkNode<T> node = this.front;
        this.front = this.front.getNext();
        node.setNext(null);

        if(this.front == null) this.back = null;
    }

    public T getFront()
    {
        if(this.front == null) return this.defaultValue;

        return this.front.getData();
    }

    public T getBack()
    {
        if(this.back == null) return this.defaultValue;

        return this.back.getData();
    }

    public bool isEmpty()
    {
        return this.front == null;
    }

    public int getSize()
    {
        return this.size;
    }

    public void clear()
    {
        while(this.front != null)
        {
            this.size--;
            SingleLinkNode<T> node = this.front;
            this.front = this.front.getNext();
            node.setNext(null);
        }

        this.back = null;
        this.front = null;
    }
}

File: data-structures/c#/Queue.cs
namespace DataStructures
{
	public class Queue<T>
	{
		private class Node
		{
			public T data;
			public Node next;

			public Node(T d, Node n = null)
			{
				data = d;
				next = n;
			}
		}

		private long count = 0;
		private Node back = null;
		private Node front = null;

		public Queue()
		{}

		~Queue()
		{
			Delete();
		}

		public void Enqueue(T data)
		{
			count++;
			Node node = new Node(data);

			if(back == null || front == null) {back = node; front = node;}
			else {back.next = node; back = node;}
		}

		public void Dequeue()
		{
			if(front == null) return;

			count--;
			Node node = front;
			front = front.next;
			node.next = null;

			if(front == null) back = null;
		}

		public T Front()
		{
			if(front == null) return default(T);

			return front.data;
		}

		public T Back()
		{
			if(back == null) return default(T);

			return back.data;
		}

		public bool Empty()
		{
			return front == null;
		}

		public long Count()
		{
			return count;
		}

		public void Delete()
		{
			while(front != null)
			{
				count--;
				Node node = front;
				front = front.next;
				node.next = null;
			}

			back = null;
			front = null;
		}
	}
}